darshan-util/pydarshan/darshan/tests/test_cffi_misc.py

```diff
@@ -159,3 +160,107 @@ def test_log_get_generic_record(dtype):
+    # pack pandas DataFrame objects back into
+    # a contiguous buffer of several records
+    # and then use the darshan-util C lib accumulator
+    # on that record array, and compare the results
+    # with those discussed in gh-867 from Perl report
+    log_path = get_log_path("imbalanced-io.darshan")
+    with darshan.DarshanReport(log_path, read_all=True) as report:
+        report.mod_read_all_records("POSIX", dtype="pandas")
+        rec_dict = report.records["POSIX"][0]
+    counters_df = rec_dict["counters"]
+    fcounters_df = rec_dict["fcounters"]
+    # gh-867 and the perl report filtered files that were
+    # only stat'd rather than opened, so demo the same filtering
+    # here at Python layer, then feed back to C accum stuff
+    fcounters_df = fcounters_df[counters_df["POSIX_OPENS"] > 0]
+    counters_df = counters_df[counters_df["POSIX_OPENS"] > 0]
+    rec_dict["counters"] = counters_df
+    rec_dict["fcounters"] = fcounters_df
+    record_array = backend._df_to_rec(rec_dict, "POSIX")
+    num_recs = rec_dict["fcounters"].shape[0]
+    mod_type = backend._structdefs["POSIX"]
+    buf = ffi.new("void **")
+    rbuf = ffi.cast(mod_type, buf)
+
+	# need to deal with the low-level C stuff to set up
+    # accumulator infrastructure to receive the repacked
+    # records
+    log_handle = backend.log_open(log_path)
+    jobrec = ffi.new("struct darshan_job *")
+    libdutil.darshan_log_get_job(log_handle['handle'], jobrec)
+    modules = backend.log_get_modules(log_handle)
+    darshan_accumulator = ffi.new("darshan_accumulator *")
+    r = libdutil.darshan_accumulator_create(modules["POSIX"]['idx'],
+                                            jobrec[0].nprocs,
+                                            darshan_accumulator)
+    assert r == 0
+    r_i = libdutil.darshan_accumulator_inject(darshan_accumulator[0], record_array, num_recs)
+    assert r_i == 0
+    darshan_derived_metrics = ffi.new("struct darshan_derived_metrics *")
+    r = libdutil.darshan_accumulator_emit(darshan_accumulator[0],
+                                          darshan_derived_metrics,
+                                          rbuf)
+    assert r == 0
+    # the indices into category_counters are pretty opaque.. we should just
+    # move everything to Python eventually... (also to avoid all the junk above when filtering..)
+    # 0 = total
+    # 1 = RO
+    # 2 = WO
+    # 3 = R/W
+    expected_total_files = 28 # see gh-867
+    expected_ro_files = 13 # see gh-867
+    expected_wo_files = 12 # see gh-867
+    actual_total_files = darshan_derived_metrics.category_counters[0].count
+    actual_ro_files = darshan_derived_metrics.category_counters[1].count
+    actual_wo_files = darshan_derived_metrics.category_counters[2].count
+    libdutil.darshan_free(buf[0])
+    backend.log_close(log_handle)
+    assert actual_total_files == expected_total_files
+    assert actual_ro_files == expected_ro_files
+    assert actual_wo_files == expected_wo_files
```

Comment:
To give an idea of how close we are, the test passes locally with no segfault issues with this diff to match `darshan-parser --show-incomplete --file` from gh-867 (but I wanted to demo the filtering to remove files that were only stat'd). Something is still sensitive when I run the entire `darshan` suite though.. probably memory management.

```diff
--- a/darshan-util/pydarshan/darshan/tests/test_cffi_misc.py
+++ b/darshan-util/pydarshan/darshan/tests/test_cffi_misc.py
@@ -220,8 +220,8 @@ def test_reverse_record_array():
     # here at Python layer, then feed back to C accum stuff
     fcounters_df = fcounters_df[counters_df["POSIX_OPENS"] > 0]
     counters_df = counters_df[counters_df["POSIX_OPENS"] > 0]
-    rec_dict["counters"] = counters_df
-    rec_dict["fcounters"] = fcounters_df
+    #rec_dict["counters"] = counters_df
+    #rec_dict["fcounters"] = fcounters_df
     record_array = backend._df_to_rec(rec_dict, "POSIX")
     num_recs = rec_dict["fcounters"].shape[0]
     mod_type = backend._structdefs["POSIX"]
@@ -253,9 +253,9 @@ def test_reverse_record_array():
     # 1 = RO
     # 2 = WO
     # 3 = R/W
-    expected_total_files = 28 # see gh-867
-    expected_ro_files = 13 # see gh-867
-    expected_wo_files = 12 # see gh-867
+    expected_total_files = 1026 # see gh-867
+    expected_ro_files = 12 # see gh-867
+    expected_wo_files = 2 # see gh-867
     actual_total_files = darshan_derived_metrics.category_counters[0].count
     actual_ro_files = darshan_derived_metrics.category_counters[1].count
     actual_wo_files = darshan_derived_metrics.category_counters[2].count
```